### ISSUE TITLE
 Monte-Carlo sampling

### ISSUE THREAD

**Issue by ZhouCX117** [Li-Wanhua/POEs#1]
Hi, Thank you for your excellent work!
The paper mentioned that  Monte-Carlo sampling was only used in the training process.
However, I find the same training and testing process in the code.
Could you help me understand this? Thanks a lot!

**Comment by xk-huang** [comment]
Hi, thanks for your attention to our work!
In both train and test codes, we calculate the MAE and Accuracy metrics with or without sampling.
https://github.com/Li-Wanhua/POEs/blob/f8efd3daa9bbb2abdaae61ed93c487d6de98968d/codes/adience_poe/train.py#L108-L110
https://github.com/Li-Wanhua/POEs/blob/f8efd3daa9bbb2abdaae61ed93c487d6de98968d/codes/adience_poe/test.py#L52-L54

**Comment by Li-Wanhua** [comment]
Note that we test both settings to evaluate the performance difference in our codes. Since we found no significant difference, we adopt the No Sampling strategy in the test phase for efficiency. You can directly drop the Sampling codes for testing.

**Comment by ZhouCX117** [comment]
Thank you for your answer, I can understand now.
However, there occurs a new issue.
When I trained on the AFAD dataset, the loss function could not converge and get collapsed in epoch 2.
![image](https://user-images.githubusercontent.com/79554332/115977391-4ecccc80-a5aa-11eb-8043-d61dc7d7f5c4.png)
Could you please give me some suggestions?
In fact, it usually happens when I train with KL loss. So it's an annoying problem！

**Comment by ZhouCX117** [comment]
And the training costs a lot of time, is it normal?
It costs about ten hours to train 1.5 epoch on a GeForce RTX 2080.

**Comment by ZhouCX117** [comment]
![image](https://user-images.githubusercontent.com/79554332/115977618-57260700-a5ac-11eb-979a-203a31308c87.png)
I can understand that your purpose is to build triples, how did you come up with this?

**Comment by Li-Wanhua** [comment]
1. Hi, for the NaN issue, we have not met it on the databases reported in our paper. But some common practices are trying different learning rates, careful of torch.log() (add a small constant, e.g., 1e-9 to avoid zero inputs) before KLloss().
2. Maybe the training data of AFAD is much larger.  For reference, the training speed of our method is ~0.4s/per batch where each batch has 32 samples. The speed is tested on a GeForce RTX 2080 Ti.
3. Here we want to build triplets with an online hard example mining strategy, more details can be found in the Supplementary Materials.

**Comment by ZhouCX117** [comment]
Thank you for your answer! It is very helpful!!
Could I ask a more question?
I want to train the model on multiple GPUs, and I use "nn.DataParallel()". It does not work. 
The shape of logit is (max_t, batchsize/num_gpus,1), which should be (max_t, batchsize,1).
Could you please give me more suggestions? Thanks a lot!

**Comment by xk-huang** [comment]
Hi, @ToBeNormal ! When we implemented the codes, we did not consider multi-GPU supporting. 
Here are two naive solutions from my personal view, for both of them you might have to modify the sampling-corresponding codes (`probordiloss.py`, `metrics.py`, `vgg.py`, and loss-computing parts in `train.py`):
1. Set max_t=1, and reshape the logit to shape (batch_size, dim_emb)
2. Expand the embedding in the second axis, i.e. (batch_size, max_t, dim_emb). The codes may look like `emb = emb.unsqueeze(1).expand(batch_size, max_t, dim_emb)`
